p2p/net/swarm/dial_sync.go

```diff
@@ -96,14 +95,21 @@ func (ds *dialSync) Dial(ctx context.Context, p peer.ID) (*Conn, error) {
 		return nil, err
 	}
 
-	defer func() {
-		ds.mutex.Lock()
-		defer ds.mutex.Unlock()
-		ad.refCnt--
-		if ad.refCnt == 0 {
-			ad.close()
-			delete(ds.dials, p)
+	conn, err := ad.dial(ctx)
+
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
+	ad.refCnt--
+	if ad.refCnt == 0 {
+		if err == nil {
+			ad.cancelCause(errConcurrentDialSuccessful)
```

Comment:
Why? What would happen if we're just dialing one address, and that dial is successful?

Reply:
That dial would have succeeded so the context cancellation has no effect on that dial. That method would have returned. 

There are two options:
1. We set call `cancelCause(nil)` and live with the brittle check that `errors.Is(Cause(err), context.Canceled)` signals concurrent successful dial. Here we override actual cancelations of parent context with errParentCanceled. This is the strategy in 055cec0c7c1e8341b2bd5dc7b31ee7ae591d9ad4 commit. 
2. Is the latest commit. We explicitly signal `errConcurrentDialSuccessful` when err is nil. In case the dial succeeds, this is not a problem since we will not call `FailedDial` for tracking metrics in that case.